File: apps/portal/web/src/api/podset.ts
import { apiFetch, buildQuery, type ListParams, type Pagination } from "./client";

export interface PodSetListItem {
  name: string;
  namespace: string;
  podGroupSize: number;
  readyPods: number;
  totalPods: number;
  phase: string;
  createdAt: string;
}

export interface PodSetListResponse {
  items: PodSetListItem[];
  pagination: Pagination;
}

export interface PodSetCreateRequest {
  name: string;
  namespace: string;
  podGroupSize: number;
  stateful?: boolean;
}

export interface PodSetDetail {
  name: string;
  namespace: string;
  podGroupSize: number;
  stateful: boolean;
  status: { readyPods: number; totalPods: number; phase: string };
  createdAt: string;
}

export const podSetApi = {
  list: (params?: ListParams) =>
    apiFetch<PodSetListResponse>(`/podsets${buildQuery(params)}`),
  get: (ns: string, name: string) =>
    apiFetch<PodSetDetail>(`/podsets/${ns}/${name}`),
  create: (data: PodSetCreateRequest) =>
    apiFetch<PodSetDetail>("/podsets", { method: "POST", body: JSON.stringify(data) }),
  delete: (ns: string, name: string) =>
    apiFetch<void>(`/podsets/${ns}/${name}`, { method: "DELETE" }),
};

File: pkg/portal/types/podset.go
package types

import "time"

type PodSetCreateRequest struct {
	Name         string `json:"name" binding:"required"`
	Namespace    string `json:"namespace" binding:"required"`
	PodGroupSize int32  `json:"podGroupSize" binding:"required"`
	Stateful     bool   `json:"stateful,omitempty"`
}

type PodSetUpdateRequest struct {
	PodGroupSize *int32 `json:"podGroupSize,omitempty"`
}

type PodSetListItem struct {
	Name         string    `json:"name"`
	Namespace    string    `json:"namespace"`
	PodGroupSize int32     `json:"podGroupSize"`
	ReadyPods    int32     `json:"readyPods"`
	TotalPods    int32     `json:"totalPods"`
	Phase        string    `json:"phase"`
	CreatedAt    time.Time `json:"createdAt"`
}

type PodSetListResponse struct {
	Items      []PodSetListItem `json:"items"`
	Pagination Pagination       `json:"pagination"`
}

type PodSetDetailResponse struct {
	Name         string             `json:"name"`
	Namespace    string             `json:"namespace"`
	PodGroupSize int32              `json:"podGroupSize"`
	Stateful     bool               `json:"stateful"`
	Status       PodSetDetailStatus `json:"status"`
	CreatedAt    time.Time          `json:"createdAt"`
}

type PodSetDetailStatus struct {
	ReadyPods int32  `json:"readyPods"`
	TotalPods int32  `json:"totalPods"`
	Phase     string `json:"phase"`
}
